flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java

```diff
@@ -152,12 +160,34 @@ class SortMergeResultPartitionReadScheduler implements Runnable, BufferRecycler
 
     @Override
     public synchronized void run() {
-        Queue<SortMergeSubpartitionReader> availableReaders = getAvailableReaders();
-
-        Queue<MemorySegment> buffers = allocateBuffers(availableReaders);
+        Set<SortMergeSubpartitionReader> finishedReaders = new HashSet<>();
+        Queue<MemorySegment> buffers;
+        try {
+            buffers = allocateBuffers();
+        } catch (Throwable throwable) {
+            // fail all pending subpartition readers immediately if any exception occurs
+            LOG.error("Failed to request buffers for data reading.", throwable);
+            failSubpartitionReaders(getAllReaders(), throwable);
```

Comment:
If a failed reader not removed from `sortedBuffers` in time, It can only be removed after the next poll, but this failed reader will read data from disk int that case, which is an unnecessary extra overhead. IMO, we need clear `sortedReaders`  when `allocateBuffers` throw exception.

Reply:
`release` and `releaseSubpartitionReader` should take care of `sortedReaders ` also. These two method will add reader to `failedReaders` set, I suggest add check to `getNextReaderToRead`, when the polled reader contains in `failedReaders`, we can ignore it and poll next until meet the reader not failed.

Reply:
@reswqa Thanks for helping review. Good point. 

In order to clear the reader in `sortedBuffers` in time, I added a check whether the reader is in `failedReaders` in the method `addPreviousAndGetNextReader`. 
In addtion, when `allReaders` is empty, the `sortedBuffers` will be cleared in `removeFinishedAndFailedReaders`.